### PR TITLE
HPCC-15618 Systemd and init updates

### PR DESCRIPTION
**Pull request by Michael-Gardner**
Signed-off-by: Michael Gardner <michael.gardner@lexisnexis.com>

### CODE REVIEW

**Comment by Michael-Gardner** [comment]
Main changes are:

- full systemd implementation.  I've created an entire set of scripts to generated systemd service files for every component.  They're capable of running standalone or as systemd instances (and are ran as systemd instances from the new hpcc-init.)
- daemon call for each component allowing them to double fork and then write out a pidfile.  Systemd prefers this when handling cgroups.  It isn't used in the old sysV init scripts so their behavior stays exactly the same.
- hpcc-init rewrite.  It now handles both sysV and systemd depending what's enabled on the system.  It handles dafilesrv as well.  The logic primarily stays the same, except that the script now uses the bash internal getopts so it will be portable to osx as well.  (The getopt version on osx tends to be different than what is available on a gnu based linux system.)

**Comment by richardkchapman** [comment]
This seems to have stalled - @mckellyln  @kunalaswani  @xwang2713 Any reviews?

**Comment by xwang2713** [comment]
I briefly reviewed the cod and will go through it again. Also did a private build and will install/play with it.

**Comment by xwang2713** [comment]
I played and briefly tested on Ubuntu 16.04 system. Here are some what I found:

1. hpcc-init usage should include "status" and "setup"
1. Call "sudo service spcc-init start" on systemed enabled system doesn't have any output on STDOUT/STDERR.  We may recommend user call hpcc-init directly since usually we don't expect spcc-init start automatically after system boot.
1. Need check /var/run/HPCCSystems first and owned by "hpcc", otherwise components will be failed to start (no pid can be created). Also make sure /var/lib/HPCCSystems and /var/log/HPCCSystems
1. dali start fails due to ""Another DASERVER process is currently running". But I checked there was no deserved. It did have daliadmin process running.

**Comment by xwang2713** [comment]
I verified "Another DASERVER process is currently running" is due to missing /var/lock/HPCCSystems directory or permission prolbem.

File: initfiles/bash/etc/init.d/install-init.in [hunk 1]
```diff
@@ -282,22 +285,11 @@ chown root:$group ${configs}
...
-      cp ${INSTALL_DIR}/etc/init.d/dafilesrv.service /lib/systemd/system/
-      cp ${INSTALL_DIR}/etc/init.d/dafilesrv@.service /lib/systemd/system/
-      systemctl daemon-reload
-   fi
-fi
+# enable systemd support if applicable
+which systemctl 1>/dev/null 2>&1 && systemctl daemon-reload
 
-createRuntime
+mkdir -p ${runtime} ${lock} ${pid} ${log}
```
Comment (Michael-Gardner) [inline initfiles/bash/etc/init.d/install-init.in]: @xwang2713 Ming this should be creating that directory on installation and the following line should insure the permissions are correct.  Can you confirm that this is in the build you're using?

**Comment by xwang2713** [comment]
It is not in the build I have. Let me rebuild with your HPCC-15618 branch. I see there are several HPCC-15618 related branch on your repo. The pull request seems links branch HPCC-15618. If not let me know

**Comment by Michael-Gardner** [comment]
Yes, The one you should be pulling is HPCC-15618.  The others named similar are branches I used to test things, or to debug on different OS's as I was trying to figure out all the edge cases where things would break.

**Comment by Michael-Gardner** [comment]
Rebased as requested.

**Comment by Michael-Gardner** [comment]
After Ming's suggestions, I've decided to close this PR and I'll be putting up a separate branch.
